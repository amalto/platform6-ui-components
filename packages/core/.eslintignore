dist/
loader/

src/components.d.ts