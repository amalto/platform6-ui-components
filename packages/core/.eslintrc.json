{
  "extends": ["../../.eslintrc.json", "plugin:@stencil/recommended", "plugin:jest/recommended", "plugin:jsx-a11y/recommended"],
  "parserOptions": {
    "project": ["./tsconfig.json", "./tsconfig.lint.json"],
    "ecmaFeatures": {
      "jsx": true // Allows for the parsing of JSX
    }
  },
  "plugins": ["jest"],

  "rules": {
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "assert": "either"
      }
    ],
    "react/jsx-props-no-spreading": "off",
    "react/no-unknown-property": [2, { "ignore": ["class"] }]
  },
  "settings": {
    "react": {
      "pragma": "h"
    }
  }
}
