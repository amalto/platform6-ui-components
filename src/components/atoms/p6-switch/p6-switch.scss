//@import "src/global/_colors";
//@import "src/global/_variable";
@import '../../../styles/atoms/control';

$switch-background: $grey-light !default;
$switch-border: 0.1rem solid $switch-background !default;
$switch-background-active: $primary !default;
$switch-radius: $radius !default;
$switch-paddle-background: $white !default;
$switch-paddle-background-active: $primary !default;
$switch-paddle-offset: 0.25rem !default;
$switch-paddle-transition: all 0.25s ease-out !default;
$switch-focus: none !default;

@mixin switch-size($size) {
  $switch-height: $size * 1.5;
  $switch-width: $switch-height * 2;
  $paddle-height: $switch-height - ($switch-paddle-offset * 2);
  $paddle-width: $switch-height - ($switch-paddle-offset * 2);
  $paddle-active-offset: $switch-width - $paddle-width - ($switch-paddle-offset * 1.5);

  + label {
    font-size: $size;
    padding-left: $switch-width + 0.5;

    &::before {
      width: $switch-width;
      height: $switch-height;
      border: $switch-border;
      background-color: transparent;
      border-radius: $radius-large * 4;
    }

    &::after {
      top: ($switch-height / 2) - ($paddle-height / 2);
      left: $switch-paddle-offset;
      width: $paddle-width;
      height: $paddle-height;
      transition: $switch-paddle-transition;
      background: $switch-background;
      border-radius: 50%;
    }
  }

  &:checked {
    + label {
      &::before {
        background: transparent;
        border-color: $switch-background-active;
      }

      &::after {
        left: $paddle-active-offset;
        background: $switch-paddle-background-active;
      }
    }
  }
}

:host {
  @include control;

  .switch[type='checkbox'] {
    vertical-align: baseline;
    font-family: $body-family;
    outline: 0;
    user-select: none;
    display: inline-block;
    position: absolute;
    opacity: 0;

    + label {
      position: relative;
      display: initial;
      line-height: initial;
      padding-top: 0.25rem;
      cursor: pointer;

      &::before {
        position: absolute;
        display: block;
        top: 0;
        left: 0;
        content: '';
        outline: none;
      }

      &::after {
        display: block;
        position: absolute;
        transform: translate3d(0, 0, 0);
        content: '';
        outline: none;
      }
    }

    &:focus {
      + label {
        &::before,
        &::after {
          outline: $switch-focus;
        }
      }
    }

    &[disabled] {
      cursor: not-allowed;

      + label {
        opacity: 0.5;

        &::before {
          opacity: 0.5;
        }

        &::after {
          opacity: 0.5;
        }

        &:hover {
          cursor: not-allowed;
        }
      }
    }

    @include switch-size($size-normal);

    @each $name, $pair in $colors {
      $color: nth($pair, 1);
      $color-invert: nth($pair, 2);
      &.is-#{$name} {
        &:checked {
          + label {
            &::before {
              background-color: transparent;
              border-color: $color !important;
            }

            &::after {
              background: $color;
            }
          }
        }
      }
      &.is-unchecked-#{$name} {
        + label {
          &::before {
            background-color: transparent;
            border-color: $color !important;
          }

          &::after {
            background: $color;
          }
        }
      }
    }
  }
}
