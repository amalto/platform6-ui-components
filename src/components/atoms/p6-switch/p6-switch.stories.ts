import { Components } from "../../../components";
import {
  getElement,
  makeStory,
  ModeStoryMaker,
  Props,
  SizeStoryMaker,
} from "../../../shared/storybook/stories";
import { Mode, Size } from "../../../shared/types";

const component = "p6-switch";

export default {
  title: "Atoms/Switch",
  component,
};

const getStoryField = (
  label: string,
  props?: Props<Components.P6Switch>
): HTMLElement => getElement(component, label, props);

export const Default = makeStory<{
  disabled: boolean;
  size: Size;
  mode: Mode;
  label: string;
}>({
  args: {
    disabled: false,
    size: Size.normal,
    mode: Mode.default,
    label: "Switch label",
  },
  builder: ({ label, ...args }): HTMLElement =>
    getStoryField(label, { ...args }),
});

export const Disabled = makeStory({
  builder: (): HTMLElement => getStoryField("Label", { disabled: true }),
});

export const Enabled = makeStory({
  builder: (): HTMLElement => getStoryField("Label", { disabled: false }),
});

export const Sizes = makeStory({
  builder: (): HTMLElement =>
    SizeStoryMaker(({ key, value }) =>
      getStoryField(key, {
        size: value,
      })
    ),
});

export const Modes = makeStory({
  builder: (): HTMLElement =>
    ModeStoryMaker(({ key, value }) =>
      getStoryField(key, {
        mode: value,
      })
    ),
});
