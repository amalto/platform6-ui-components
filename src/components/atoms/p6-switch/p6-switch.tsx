import {
  Component,
  ComponentInterface,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Method,
  Prop,
  State,
} from "@stencil/core";
import { P6Control } from "~shared/form/control";
import { InvalidEventDetail, ValidEventDetail } from "~shared/form/event";
import { Mode, Size } from "~shared/types";
import { getModeClass, getSizeClass } from "~utils/classes";

export type P6SwitchValue = boolean | undefined;

@Component({
  tag: "p6-switch",
  styleUrl: "./p6-switch.scss",
  shadow: true,
})
export class P6Switch implements ComponentInterface, P6Control<P6SwitchValue> {
  @Element() host!: HTMLP6SwitchElement;

  /**
   * Initial value
   */
  @Prop() checked = false;

  /**
   * Disable
   */
  @Prop() disabled = false;

  /**
   * Mode
   */
  @Prop() color: Mode = Mode.success;

  /**
   * Size
   */
  @Prop() size: Size = Size.normal;

  /**
   * Switch name
   */
  @Prop() name!: string;

  /**
   * Registering the field in a p6-form
   */
  @Event() p6FormRegister!: EventEmitter<P6Control<P6SwitchValue>>;

  /**
   * Unregistering the field in a p6-form
   */
  @Event() p6FormUnregister!: EventEmitter<P6Control<P6SwitchValue>>;

  /**
   * Fires when the field has been checked for validity and satisfy its constraints
   */
  @Event() p6Valid!: EventEmitter<ValidEventDetail<P6SwitchValue>>;

  /**
   * Fires when the field has been checked for validity and doesn't satisfy its constraints
   */
  @Event() p6Invalid!: EventEmitter<InvalidEventDetail>;

  @State() hasError = false;

  private nativeInput: HTMLInputElement | undefined;

  // private defaultValue: boolean = this.checked;

  componentWillLoad(): void {
    // this.defaultValue = this.checked;
    this.host.addEventListener("focusout", this.checkValidity.bind(this));
  }

  render(): JSX.Element {
    const { name, checked, disabled } = this;

    const inputId = `${name}-input`;

    return (
      <Host onClick={this.onClick}>
        <input
          class={{
            switch: true,
            ...getSizeClass(this.size),
            ...getModeClass(this.color),
          }}
          ref={(input): void => {
            this.nativeInput = input;
          }}
          checked={checked}
          disabled={disabled}
          id={inputId}
          name={name}
          type="checkbox"
        />
        <label htmlFor={inputId}>
          <slot />
        </label>
      </Host>
    );
  }

  componentDidLoad(): void {
    this.p6FormRegister.emit(this);
  }

  disconnectedCallback?(): void {
    this.p6FormUnregister.emit(this);
  }

  /*
  @Watch("checked")
  onCheckedUpdated(newValue: boolean) {
    this.defaultValue = newValue;
  }
*/

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    const message = this.nativeInput?.validationMessage || "";
    return Promise.resolve(message);
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    const isValid = !!this.nativeInput?.checkValidity();

    const message = await this.validationMessage();
    this.hasError = message !== "";

    if (isValid) {
      const validInit = {
        name: this.name,
        value: this.checked,
      };
      this.p6Valid.emit(validInit);
    } else {
      const invalidInit = {
        name: this.name,
        message,
      };
      this.p6Invalid.emit(invalidInit);
    }

    return Promise.resolve(isValid);
  }

  private onClick = (): void => {
    this.checked = !this.checked;
  };
}
