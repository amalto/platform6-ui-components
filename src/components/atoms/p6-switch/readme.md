# p6-switch

<!-- Auto Generated Below -->


## Properties

| Property            | Attribute  | Description   | Type                                                                                       | Default        |
| ------------------- | ---------- | ------------- | ------------------------------------------------------------------------------------------ | -------------- |
| `checked`           | `checked`  | Initial value | `boolean`                                                                                  | `false`        |
| `color`             | `color`    | Mode          | `Mode.danger \| Mode.default \| Mode.info \| Mode.primary \| Mode.success \| Mode.warning` | `Mode.success` |
| `disabled`          | `disabled` | Disable       | `boolean`                                                                                  | `false`        |
| `name` _(required)_ | `name`     | Switch name   | `string`                                                                                   | `undefined`    |
| `size`              | `size`     | Size          | `Size.large \| Size.medium \| Size.normal \| Size.small`                                   | `Size.normal`  |

## Events

| Event              | Description                                                                            | Type                                             |
| ------------------ | -------------------------------------------------------------------------------------- | ------------------------------------------------ |
| `p6FormRegister`   | Registering the field in a p6-form                                                     | `CustomEvent<P6Control<P6CheckboxValue>>`        |
| `p6FormUnregister` | Unregistering the field in a p6-form                                                   | `CustomEvent<P6Control<P6CheckboxValue>>`        |
| `p6Invalid`        | Fires when the field has been checked for validity and doesn't satisfy its constraints | `CustomEvent<InvalidEventDetail>`                |
| `p6Valid`          | Fires when the field has been checked for validity and satisfy its constraints         | `CustomEvent<ValidEventDetail<P6CheckboxValue>>` |

## Methods

### `checkValidity() => Promise<boolean>`

Returns whether a form will validate when it is submitted, without having to submit it.

#### Returns

Type: `Promise<boolean>`

### `validationMessage() => Promise<string>`

Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
It also triggers the standard error message, such as "this is a required field".

#### Returns

Type: `Promise<string>`


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
