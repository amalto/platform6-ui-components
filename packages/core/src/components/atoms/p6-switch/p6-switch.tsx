import { Component, ComponentInterface, Element, Event, EventEmitter, h, Host, JSX, Method, Prop, State } from '@stencil/core';
import { P6Control } from '../../../shared/form/control';
import { InvalidEventDetail, ValidEventDetail } from '../../../shared/form/event';
import { defaultCheckValidity, defaultValidationMessage } from '../../../shared/form/validation';
import { Mode } from '../../../shared/types';
import { getModeClass } from '../../../utils/classes';

export type P6SwitchValue = boolean | undefined;

@Component({
  tag: 'p6-switch',
  styleUrl: './p6-switch.scss',
  shadow: true,
})
export class P6Switch implements ComponentInterface, P6Control<P6SwitchValue> {
  @Element() host!: HTMLP6SwitchElement;

  /**
   * Initial value
   */
  @Prop() checked = false;

  /**
   * Disable
   */
  @Prop() disabled = false;

  /**
   * Mode
   */
  @Prop() mode: Mode = Mode.success;

  /**
   * Switch name
   */
  @Prop() name!: string;

  /**
   * Registering the field in a p6-form
   */
  @Event() p6FormRegister!: EventEmitter<P6Control<P6SwitchValue>>;

  /**
   * Unregistering the field in a p6-form
   */
  @Event() p6FormUnregister!: EventEmitter<P6Control<P6SwitchValue>>;

  /**
   * Fires when the field has been checked for validity and satisfy its constraints
   */
  @Event() p6Valid!: EventEmitter<ValidEventDetail<P6SwitchValue>>;

  /**
   * Fires when the field has been checked for validity and doesn't satisfy its constraints
   */
  @Event() p6Invalid!: EventEmitter<InvalidEventDetail>;

  @State() hasError = false;

  private nativeInput: HTMLInputElement | undefined;

  componentWillLoad(): void {
    this.host.addEventListener('focusout', this.checkValidity.bind(this));
  }

  render(): JSX.Element {
    const { name, checked, disabled } = this;

    const inputId = `${name}-input`;

    return (
      <Host>
        <input
          class={{
            switch: true,
            ...getModeClass(this.mode),
          }}
          ref={(input): void => {
            this.nativeInput = input;
          }}
          checked={checked}
          disabled={disabled}
          id={inputId}
          name={name}
          onClick={this.onClick}
          type="checkbox"
        />
        <label htmlFor={inputId}>
          <slot />
        </label>
      </Host>
    );
  }

  componentDidLoad(): void {
    this.p6FormRegister.emit(this);
  }

  disconnectedCallback?(): void {
    this.p6FormUnregister.emit(this);
  }

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    return defaultValidationMessage(this.nativeInput);
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    return defaultCheckValidity<P6SwitchValue>({
      name: this.name,
      disabled: this.disabled,
      nativeInput: this.nativeInput,
      p6Valid: this.p6Valid,
      p6Invalid: this.p6Invalid,
      validationMessage: this.validationMessage.bind(this),
      errorHandler: hasError => {
        this.hasError = hasError;
      },
      getValue: () => this.checked,
    });
  }

  private onClick = (): void => {
    this.checked = !this.checked;
  };
}
