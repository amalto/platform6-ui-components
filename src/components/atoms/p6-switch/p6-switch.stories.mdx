import { Story, Preview, Meta } from "@storybook/addon-docs/blocks";
import { action, decorate } from "@storybook/addon-actions";
import { select, boolean, text, withKnobs } from "@storybook/addon-knobs";
import readme from "./readme.md";

<Meta title="Atoms/Switch" parameters={{ notes: readme }} />

# Switch

`p6-switch` provide a switch component which allow you to _enable_ or _disable_ a field.

<Story
  name="Default"
  parameters={{
    docs: { disable: true },
    actions: {
      handles: ["p6Submit p6-form"],
    },
  }}
  decorators={[withKnobs]}
>
  {() => {
    const kDisabled = boolean("Disabled", false);
    const kColors = select(
      "Mode",
      ["danger", "warning", "default", "info", "success", "primary"],
      "success"
    );
    const kSizes = select(
      "Size",
      ["small", "normal", "medium", "large"],
      "normal"
    );
    const kText = text("Text", "Switch on by default");
    return `
      <p6-form>
        <p6-switch name="test"
          disabled=${kDisabled}
          mode=${kColors}
          size=${kSizes}
          >${kText}</p6-switch>
          <p6-button class="button is-link">Submit</p6-button>
          <p6-button class="button is-link" type="reset">Reset</p6-button>
      </p6-form>
    `;
  }}
</Story>

### Enabled

Basic implementation enabled by default.

<Preview>
  <Story name="Enabled" height="44px">
    {`
    <p6-switch checked name="switch-enabled">
      Default switch
    </p6-switch>
  `}
  </Story>
</Preview>

### Disabled

The `disabled` _prop_ will keep you from editing it. In the case you are using it inside a form, it won't populate the submit object.

<Preview>
  <Story name="Disabled" height="44px">
    {`
    <p6-switch checked disabled name="switch-disabled">
      Disabled switch
    </p6-switch>
  `}
  </Story>
</Preview>

### Colors

<Preview>
  <Story name="Colors" height="44px">
    {["danger", "warning", "default", "info", "success", "primary"]
      .map(
        (color) =>
          `<p><p6-switch checked color="${color}">${color}</p6-switch></p>`
      )
      .join("")}
  </Story>
</Preview>

### Sizes

<Preview>
  <Story name="Sizes" height="44px">
    {["small", "normal", "medium", "large"]
      .map((size) => `<p><p6-switch size="${size}">${size}</p6-switch></p>`)
      .join("")}
  </Story>
</Preview>
