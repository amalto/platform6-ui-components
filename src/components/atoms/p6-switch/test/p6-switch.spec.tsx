import { newSpecPage } from '@stencil/core/testing';
import { P6Switch } from '../p6-switch';

describe('p6-switch', () => {

  const id: string = "0123456789";

  it('with label only', async () => {
    const page = await newSpecPage({
      components: [P6Switch],
      html: `<p6-switch name="${id}">Hello world!</p6-switch>`,
    });
    expect(page.root).toEqualHtml(`
      <p6-switch name="${id}">
        <input id="${id}-input" name="${id}" type="switch" />
        <label htmlFor="${id}-input">Hello world!</label>
      </p6-switch>
    `);
  });

  it('disabled switch', async () => {
    const page = await newSpecPage({
      components: [P6Switch],
      html: `
      <p6-switch disabled="true" name="${id}">Disabled switch</p6-switch>
      `,
    });
    expect(page.root).toEqualHtml(`
      <p6-switch disabled="true" name="${id}">
        <input id="${id}-input" disabled="" name="${id}" type="switch" />
        <label htmlFor="${id}-input">Disabled switch</label>
      </p6-switch>
    `);
  });
});
