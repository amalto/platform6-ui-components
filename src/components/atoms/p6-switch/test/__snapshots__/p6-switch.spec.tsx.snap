// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-switch disabled switch 1`] = `
<p6-switch disabled="true" name="0123456789">
  <mock:shadow-root>
    <input class="is-success switch" disabled="" id="0123456789-input" name="0123456789" type="checkbox">
    <label htmlfor="0123456789-input">
      <slot></slot>
    </label>
  </mock:shadow-root>
  Disabled switch
</p6-switch>
`;

exports[`p6-switch with label only 1`] = `
<p6-switch name="0123456789">
  <mock:shadow-root>
    <input class="is-success switch" id="0123456789-input" name="0123456789" type="checkbox">
    <label htmlfor="0123456789-input">
      <slot></slot>
    </label>
  </mock:shadow-root>
  Hello world!
</p6-switch>
`;
